cmd/stack/proxy.go

```diff
@@ -0,0 +1,319 @@
+			time.Now().Format(time.RFC3339),
+			req.Method,
+			sourceURL.String(),
+			targetURL.String())
+	}
+
+	proxy.ErrorHandler = func(w http.ResponseWriter, r *http.Request, err error) {
+		targetURL := &url.URL{
+			Scheme:   stackBaseURL.Scheme,
+			Host:     stackBaseURL.Host,
+			Path:     r.URL.Path,
+			RawQuery: r.URL.RawQuery,
+		}
+
+		fmt.Fprintf(cmd.ErrOrStderr(), "[%s] ERROR proxying request to %s: %v\n",
+			time.Now().Format(time.RFC3339),
+			targetURL.String(),
+			err)
+
+		w.Header().Set("Content-Type", "text/plain")
+		w.WriteHeader(http.StatusBadGateway)
+		errorMsg := fmt.Sprintf("Proxy Error: %v\nTarget URL: %s",
+			err,
+			targetURL.String())
+		io.WriteString(w, errorMsg)
+	}
+
+	transport := &tokenTransport{
+		wrapped:  http.DefaultTransport,
+		profile:  c.profile,
+		stack:    stack,
+		cmd:      cmd,
+		tokenMux: &sync.RWMutex{},
+	}
+
+	proxy.Transport = transport
+
+	server := &http.Server{
+		Addr: fmt.Sprintf(":%d", port),
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			proxy.ServeHTTP(w, r)
+		}),
+		ReadHeaderTimeout: 10 * time.Second,
+	}
+
+	stop := make(chan os.Signal, 1)
+	signal.Notify(stop, os.Interrupt, syscall.SIGTERM)
+
+	go func() {
+		fmt.Fprintf(cmd.OutOrStdout(), "Starting proxy server at %s -> %s\n", c.store.ProxyUrl, c.store.StackUrl)
+		fmt.Fprintf(cmd.OutOrStdout(), "Press Ctrl+C to stop the server\n")
+
+		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			fmt.Fprintf(cmd.ErrOrStderr(), "Server error: %v\n", err)
+			os.Exit(1)
+		}
+	}()
+
+	<-stop
+	fmt.Fprintln(cmd.OutOrStdout(), "\nShutting down server...")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	if err := server.Shutdown(ctx); err != nil {
+		fmt.Fprintf(cmd.ErrOrStderr(), "Error during server shutdown: %v\n", err)
+	} else {
+		fmt.Fprintln(cmd.OutOrStdout(), "Server stopped successfully")
+	}
+
+	// Create a dummy renderable to avoid nil pointer dereference
+	return &EmptyRenderable{}, nil
+}
+
+// EmptyRenderable is a dummy implementation of the Renderable interface
+type EmptyRenderable struct{}
+
+func (r *EmptyRenderable) Render(cmd *cobra.Command, args []string) error {
+	return nil
+}
+
+type tokenTransport struct {
+	wrapped     http.RoundTripper
+	profile     *fctl.Profile
+	stack       *membershipclient.Stack
+	cmd         *cobra.Command
+	token       *oauth2.Token
+	tokenMux    *sync.RWMutex
+	tokenExpiry time.Time
+}
+
+func (t *tokenTransport) isTokenValid() bool {
+	// First, grab a read lock to check for nil token and expiry zero-value.
+	t.tokenMux.RLock()
+	if t.token == nil {
+		t.tokenMux.RUnlock()
+		return false
+	}
+	zero := t.tokenExpiry.IsZero()
+	t.tokenMux.RUnlock()
+
+	// If the expiry hasn't been initialized, switch to a write lock and do it.
+	if zero {
+		t.tokenMux.Lock()
+		if t.tokenExpiry.IsZero() { // double-check under write lock
+			parser := jwt.Parser{}
```

Comment:
You don't need to parse the jwt token, you can use the expires_in field of the oauth2 standard.